### PR TITLE
Inline add_integration workflow inside chat

### PR DESCRIPTION
**Pull request by azizmejri1**

### CODE REVIEW

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

File: src/components/chat/DyadAddIntegration.tsx [hunk 1]
```diff
@@ -148,7 +147,69 @@ export const DyadAddIntegration: React.FC<DyadAddIntegrationProps> = ({
...
+      requestId: pendingIntegration.requestId,
+      provider: completedProvider,
+      completed: true,
+    });
+    clearPendingForChat();
+    // The current stream ends shortly after respondToIntegration resolves the
+    // backend wait (the agent's stopWhen fires on the add_integration tool).
+    // Defer the continuation message until that stream finishes — the effect
+    // below fires it when isStreaming transitions to false.
+    setPendingContinuationProvider(completedProvider);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/chat/DyadAddIntegration.tsx]: Resolving as outdated: the specific code referenced has been refactored — `planClient.respondToIntegration` is no longer called from this component, it lives in `ConfigurePanel.tsx` now. The related "await respondToIntegration" concern has been addressed in the parallel thread on `ConfigurePanel.tsx` (try/catch + toast on failure). The broader stream-cancel cleanup would require backend coordination and is out of scope for this PR. Per **Principle #4: Transparent Over Magical**, surfacing IPC errors via toast is the highest-leverage transparency improvement for the renderer side here.

File: src/components/chat/DyadAddIntegration.tsx [hunk 2]
```diff
@@ -208,85 +272,143 @@ export const DyadAddIntegration: React.FC<DyadAddIntegrationProps> = ({
...
-          {effectiveSelectedProvider
-            ? t("integrations.databaseSetup.setUpProvider", {
-                provider:
-                  availableProviders.find(
-                    (option) => option.id === effectiveSelectedProvider,
-                  )?.name ?? t("integrations.databaseSetup.setUpDatabase"),
-              })
-            : t("integrations.databaseSetup.setUpDatabase")}
-        </Button>
+              )}
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/chat/DyadAddIntegration.tsx]: Resolving as outdated: this referenced an earlier interactive view that no longer exists. The current component does not render an integration label/skeleton in the interactive selection view — that label is only shown in the *completed* view (lines 244-249), where it already uses `h-6 w-28` and the green `bg-green-100/40 text-green-800` styling. There is no inconsistency to reconcile.

File: src/components/chat/DyadAddIntegration.tsx [hunk 3]
```diff
@@ -148,7 +147,69 @@ export const DyadAddIntegration: React.FC<DyadAddIntegrationProps> = ({
...
+        return;
+      }
+      setIsDisconnecting(false);
+    }
+    setStepOverride("select");
+  };
+
+  const handleContinueClick = () => {
+    if (!pendingIntegration || !completedProvider || chatId == null) return;
+    planClient.respondToIntegration({
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/chat/DyadAddIntegration.tsx]: Resolving as outdated: the `respondToIntegration` call has been moved out of this component into `ConfigurePanel.tsx` since this comment was written. The same concern was raised on the new location (thread on `ConfigurePanel.tsx:350`) and has been addressed there: `handleContinueClick` is now `async`, awaits the IPC call inside a try/catch, surfaces an error toast, and only proceeds with state cleanup on success. Per **Principle #4: Transparent Over Magical**, IPC failures are now visibly reported instead of silently leaving the backend hanging.

File: src/components/chat/DyadAddIntegration.tsx [hunk 4]
```diff
@@ -148,7 +147,69 @@ export const DyadAddIntegration: React.FC<DyadAddIntegrationProps> = ({
...
+    if (chatId == null) return;
+    setPendingIntegrationMap((prev) => {
+      if (!prev.has(chatId)) return prev;
+      const next = new Map(prev);
+      next.delete(chatId);
+      return next;
+    });
+  };
+
+  const handleBackClick = async () => {
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/chat/DyadAddIntegration.tsx]: Resolving as outdated: `handleBackClick` no longer calls `unsetAppProject`. The current implementation only resets local UI state — it clears `inPanelMode`, the user's provider selection, and drops the `provider` field from the pending integration so the radios reopen. The OAuth/project connection on the app itself is preserved. Disconnecting the project is now an explicit action via the connector panel, not a side effect of the Back button.

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by azizmejri1** [comment]
## 🤖 Claude Code Review Summary

### PR Confidence: 4/5
All trusted-author review threads addressed (5 valid issues fixed with code changes; 4 outdated threads resolved with explanations). One pre-existing test mock gap was discovered and fixed in the same push, so test suite is green.

### Unresolved Threads
_No unresolved threads_

### Resolved Threads
| Issue | Rationale | Link |
|-------|-----------|------|
| `selectedProvider` may not sync with async `pendingIntegration` | Replaced `useState` initializer with a derived value (`userSelectedProvider ?? requestedProvider ?? pendingIntegration?.provider ?? "supabase"`) so a late-arriving IPC `pendingIntegration?.provider` is reflected without a sync effect; user choice still wins | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189639874) |
| Keyboard arrow nav bypasses `disableSwitch` check | Added an early-return at the top of `handleRadioKeyDown` mirroring the click handler's gate (`!pendingIntegration \|\| availableProviders.length === 1`) | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3195232946) |
| Disabled radio buttons still show hover affordance | Conditionally omit `hover:border-blue-400` when `disableSwitch` is true | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189639882) |
| `planClient.respondToIntegration()` not awaited (ConfigurePanel) | Made `handleContinueClick` async, await IPC inside try/catch, surface error toast, and only proceed with state cleanup on success — per **Principle #4: Transparent Over Magical**, IPC failures are now visibly reported instead of silently leaving the backend hanging up to its 30-min timeout | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3195232941) |
| Disabled Continue button gives no explanation | Added helper text under the button when `!canContinue`: "Complete the {provider} connection above to continue." Added i18n key `completePrompt` to en/pt-BR/zh-CN. Per **Principle #4: Transparent Over Magical** | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3195232950) |
| Stale integration requests can restart canceled streams (outdated) | The specific code referenced has been refactored — the IPC call is no longer in this component. The narrower "await respondToIntegration" concern is addressed in the ConfigurePanel thread. Broader stream-cancel cleanup would need backend coordination, out of scope for this PR | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189516850) |
| Skeleton/label styling inconsistency in interactive view (outdated) | The interactive view no longer renders an integration label/skeleton; the label only appears in the completed view, where styling is already consistent (`h-6 w-28`, green palette) | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189542192) |
| `respondToIntegration` not awaited (DyadAddIntegration) (outdated) | The call has been moved out of `DyadAddIntegration` to `ConfigurePanel.tsx`; same concern addressed in the parallel ConfigurePanel thread with try/catch + toast | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189639870) |
| Back button disconnects project without confirmation (outdated) | `handleBackClick` no longer calls `unsetAppProject` — it only resets local UI state (`inPanelMode`, user provider selection, drops `provider` from pending integration). The OAuth/project link is preserved | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3189639876) |

<details>
<summary>Product Principle Suggestions</summary>

_No suggestions_ — Principle #4 (Transparent Over Magical) provided clear guidance for both the IPC error-surfacing decision and the disabled-button-helper-text decision.

</details>

---
🤖 Generated by Claude Code

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by azizmejri1** [comment]
## 🤖 Claude Code Review Summary

### PR Confidence: 4/5
All 5 unresolved review threads were valid issues addressed with code changes; remaining risk is verifying the updated E2E snapshots pass in CI for the changed `add_integration` description.

### Unresolved Threads
_No unresolved threads_

### Resolved Threads
| Issue | Rationale | Link |
|-------|-----------|------|
| Update request snapshots for changed tool description | Updated the `add_integration` tool description in 3 affected snapshots (`local_agent_basic`, `local_agent_auto`, `local_agent_advanced`) so the dump-request tests stay green. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3195850932) |
| Frontend `pendingIntegrationAtom` not cleaned up on stream cancel/abort (two related threads) | Per **Principle #4: Transparent Over Magical** — the frontend now clears `pendingIntegrationAtom` for the chat in the existing `onChatStreamEnd` listener in `renderer.tsx`, alongside the consent and questionnaire cleanups. Prevents the stale card from arming a Continue click against a backend resolver that no longer exists. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3195850938) / [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3196015980) |
| Continue button lacks loading/submitting state during async IPC call | Per **Principle #6: Delightful** — added an `isSubmitting` state with a spinner; reuses the existing `databaseSetup.continuing` i18n key, which also addresses the related dead-code thread for that key. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3196015992) |
| Dead i18n keys after refactor | Removed `setUpProvider` and `setUpDatabase` from `en/pt-BR/zh-CN` locale files. Kept `continuing` since it's now reused by the Continue button's submitting state (see thread above). | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3196015997) |

---
🤖 Generated by Claude Code

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

File: src/components/chat/DyadAddIntegration.tsx [hunk 5]
```diff
@@ -148,7 +173,72 @@ export const DyadAddIntegration: React.FC<DyadAddIntegrationProps> = ({
...
+    // Drop the chosen provider from the pending integration so the Configure
+    // panel collapses and the radios reopen with no preselection. (The
+    // tool-locked provider, if any, is still preserved via `requestedProvider`
+    // and continues to constrain `availableProviders`.)
+    if (chatId != null) {
+      setPendingIntegrationMap((prev) => {
+        const existing = prev.get(chatId);
+        if (!existing || existing.provider === undefined) return prev;
+        const next = new Map(prev);
+        next.set(chatId, { ...existing, provider: undefined });
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/chat/DyadAddIntegration.tsx:206]: Per **Principle #4: Transparent Over Magical**, the existing flow already preserves a tool-locked provider correctly. `addIntegrationTool.buildXml` always emits the `provider` attribute when `args.provider` is set, so `requestedProvider` (parsed from XML) carries the lock independently of `pendingIntegrationAtom`. `handleBackClick` only clears `pendingIntegration.provider` (the user-applied selection from clicking Next); `availableProviders` still derives from `lockedProvider = requestedProvider ?? pendingIntegration?.provider`, so a tool-locked request stays locked to a single radio after Back. The inline comment already calls this out: "The tool-locked provider, if any, is still preserved via `requestedProvider` and continues to constrain `availableProviders`." Resolving as not valid.

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by azizmejri1** [comment]
## 🤖 Claude Code Review Summary

### PR Confidence: 4/5
All eight outstanding review threads have been addressed (seven with code changes, one resolved with a code-supported counter-explanation), lint and type-check pass; the only unverified piece is end-to-end behavior in a running app.

### Unresolved Threads
_No unresolved threads_

### Resolved Threads
| Issue | Rationale | Link |
|-------|-----------|------|
| `waitForIntegrationResponse` blocks without an abort path (HIGH/P1, raised twice — `dyad-assistant` and `cubic-dev-ai`) | Threaded `abortSignal` through `AgentContext`, taught `waitForIntegrationResponse` to race against it, and updated `addIntegrationTool` to pass `ctx.abortSignal`. Cancellation now resolves the wait immediately instead of pinning the stream to the 30-min timeout. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197732767) / [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197753524) |
| Reset integration submit state after successful continue | Added a `useEffect` keyed on `pendingIntegration?.requestId` that resets `isSubmitting` whenever the active request changes or clears, so the next request's Continue button is never stuck disabled. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197569611) |
| Timed-out integration requests stay interactive | New root-level `useIntegrationContinuation` hook detects the streaming → not-streaming transition and drops any leftover `pendingIntegrationAtom` entry — covers timeout, abort, and natural completion paths uniformly. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197709964) |
| Move continuation dispatch out of the virtualized card + clear the continuation flag when the stream ends (`chatgpt-codex-connector` and `cubic-dev-ai`) | The auto-send effect was lifted from `DyadAddIntegration` (which lives inside the virtualized message list) into the new root-level `useIntegrationContinuation` hook. The hook also clears stale continuation entries on stream end, so navigation/scroll-out can no longer drop or replay the "Continue. I have completed the X integration." message. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197737773) / [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197753538) |
| Make integration notification persistent (`requireInteraction: true`) | Per **Principle #4: Transparent Over Magical**, blocking consent notifications should remain visible until acted upon. Added `requireInteraction: true` to the `Notification` constructor in `usePlanEvents`. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197753528) |
| Back clears request-provided provider state | Resolved as not valid. Per **Principle #4: Transparent Over Magical**, the existing flow correctly preserves a tool-locked provider via `requestedProvider` (XML attr), independent of `pendingIntegrationAtom`. `handleBackClick` only clears the user-applied selection; `availableProviders` still constrains to the locked provider after Back. | [View](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3197753527) |

<details>
<summary>Product Principle Suggestions</summary>

_No suggestions_

</details>

---
🤖 Generated by Claude Code

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

File: src/components/preview_panel/ConfigurePanel.tsx [hunk 6]
```diff
@@ -298,6 +311,137 @@ const EnvironmentVariablesTitle = () => (
...
+
+  const handleContinueClick = async () => {
+    if (chatId == null || !canContinue || isSubmitting) return;
+    setIsSubmitting(true);
+    // Await the IPC: if it fails (e.g. webContents destroyed during nav, or a
+    // serialization error) the backend's waitForIntegrationResponse promise
+    // would otherwise hang to its 30-min timeout while the UI moves on as if
+    // everything succeeded. On error, surface a toast and leave state intact.
+    try {
+      await planClient.respondToIntegration({
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/preview_panel/ConfigurePanel.tsx]: Addressed in 8036bc13: the Continue handler now queues the continuation in `pendingContinuationProviderAtom` **before** awaiting `respondToIntegration`, and rolls back on IPC failure. This guarantees the entry exists by the time `useIntegrationContinuation` observes the streaming → not-streaming transition, so the hidden `Continue. I have completed the X integration.` message can no longer be lost.

Per **Principle #4: Transparent Over Magical**, the rollback path also surfaces IPC failures via toast rather than leaving stale state.

[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/components/preview_panel/ConfigurePanel.tsx]: Addressed in 8036bc13: same fix as the duplicate MEDIUM thread above. The Continue handler now writes `pendingContinuationProviderAtom` **before** awaiting `respondToIntegration`, with a rollback on IPC failure, so the streaming → not-streaming transition can't be consumed before the continuation is recorded.

Per **Principle #4: Transparent Over Magical**, IPC failures surface via toast and state is left intact for the user to retry.

File: src/ipc/handlers/plan_handlers.ts [hunk 7]
```diff
@@ -165,4 +168,13 @@ export function registerPlanHandlers() {
       resolveQuestionnaireResponse(params.requestId, params.answers);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/plan_handlers.ts]: Addressed in 8036bc13: `resolveIntegrationResponse` now returns a boolean indicating whether a pending resolver was matched, and the IPC handler in `plan_handlers.ts` throws `DyadError(NotFound)` when no pending request exists for the given `requestId`. Stale, timed-out, or already-resolved requests now reject the IPC instead of silently returning success — the renderer's existing `try/catch` in `ConfigurePanel.handleContinueClick` rolls back the queued continuation and shows a toast with the error message.

Per **Principle #4: Transparent Over Magical**, classifying as `NotFound` keeps the error out of telemetry noise (it's filtered) while still surfacing it to the user.

File: src/ipc/handlers/plan_handlers.ts [hunk 8]
```diff
@@ -165,4 +168,13 @@ export function registerPlanHandlers() {
...
     },
   );
+
+  createTypedHandler(planContracts.respondToIntegration, async (_, params) => {
+    resolveIntegrationResponse(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/plan_handlers.ts]: Addressed in 8036bc13: same fix as the duplicate MEDIUM thread above. `resolveIntegrationResponse` now reports whether it matched, and the IPC handler throws `DyadError(NotFound)` for stale request IDs. The renderer's `try/catch` rolls back the queued continuation and shows the error in a toast.

Per **Principle #4: Transparent Over Magical**, the user now gets actionable feedback instead of a silent dead flow.

**Comment by azizmejri1** [comment]
## 🤖 Claude Code Review Summary

### PR Confidence: 4/5
All trusted-author review threads have been addressed with code changes; the remaining unknown is whether the integration race fix interacts cleanly with edge cases that aren't covered by existing tests.

### Unresolved Threads
_No unresolved threads_

### Resolved Threads
| Issue | Rationale | Link |
|-------|-----------|------|
| Continuation can be missed/queued too late after integration response (HIGH + MEDIUM duplicates) | Queued `pendingContinuationProviderAtom` **before** awaiting `respondToIntegration` so the streaming → not-streaming transition can't be consumed before the entry exists. Roll back on IPC failure. Per **Principle #4: Transparent Over Magical**, the rollback path surfaces IPC failures via toast. | [HIGH](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3203756734) / [MEDIUM](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3202445235) |
| Stale integration responses report success (MEDIUM duplicates) | `resolveIntegrationResponse` now returns whether it matched a pending resolver; the IPC handler throws `DyadError(NotFound)` for stale request IDs so the renderer's existing `try/catch` shows an actionable toast instead of silently treating the IPC as success. Per **Principle #4: Transparent Over Magical**. | [#1](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3203699502) / [#2](https://github.com/dyad-sh/dyad/pull/3349#discussion_r3203756745) |

<details>
<summary>Product Principle Suggestions</summary>

_No suggestions_

</details>

---
🤖 Generated by Claude Code

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Review by wwwillchen — CHANGES_REQUESTED** [review]
definitely a big ux improvement overall!

i think we should always be able to see the integration section in the configure panel. otherwise, it's confusing for users that we showed them it's there during the setup process, but if they want to change the db branch, for example, it's not there anymore.

another thing that i found a bit unintuitive was that the Continue button is in the configure panel, but i was expecting to see it on the left-side Chat area. Could we also have a continue button there? basically above the "Back" button or perhaps replacing the Back button.
Image

**Comment by azizmejri1** [comment]
> definitely a big ux improvement overall!
> 
> i think we should always be able to see the integration section in the configure panel. otherwise, it's confusing for users that we showed them it's there during the setup process, but if they want to change the db branch, for example, it's not there anymore.
> 


Yeah , i agree, hiding the integration section after the integration is done defeats the whole purpose of continuing the integration inside the config panel , so this was a mistake

**Comment by wwwillchen** [comment]
@BugBot run

**Review by wwwillchen — APPROVED** [review]
thanks! overall looks good. before merging this, can you move Integration section to the top of the Configure panel? i think in terms of importance, it's the highest, since it determines environment variable and app commands is a relatively obscure setting that's rarely changed.

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run

**Comment by wwwillchen** [comment]
@BugBot run
